cmd/relay-metrics/main.go

```diff
@@ -2,281 +2,118 @@ package main
 		Scheme: Scheme,
 	})
 	if err != nil {
-		panic(err.Error())
+		klog.Fatal(err.Error())
 	}
 
-	return c
-}
-
-func metricsClient(ctx context.Context) *monitoring.MetricClient {
-	mc, err := monitoring.NewMetricClient(ctx)
+	err = reconciler.Add(mgr, meter)
 	if err != nil {
-		panic(err.Error())
-	}
-
-	return mc
-}
-
-func main() {
-	ctx := context.Background()
-	var err error
-	var kubeconfig *string
-
-	if home := homeDir(); home != "" {
-		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
-	} else {
-		kubeconfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
-	}
-	publish := flag.Bool("publish", false, "Publish discovered metrics to stackdriver.")
-	environment := flag.String("environment", "dev", "The environment being monitored for workflow runs")
-	// Re-running the process will restore the descriptors, and if done quickly may not actually lose the data.
-	deleteMetrics := flag.Bool("delete-descriptors", false, "Delete the metric descriptors. Be very sure, there is no confirmation.")
-
-	flag.Parse()
-
-	kc := k8sClient(kubeconfig)
-	mc := metricsClient(ctx)
-
-	statusName := "Pending Workflow Runs"
-	oldestName := "Oldest Pending Workflow Run"
-	statusType := "custom.googleapis.com/relay/workflowruns/status"
-	oldestType := "custom.googleapis.com/relay/workflowruns/oldest_pending"
-
-	if *deleteMetrics == true {
-		err = deleteMetric(os.Stdout, "projects/"+projectID+"/metricDescriptors/"+statusType)
-		if err != nil {
-			panic(err.Error())
-		}
-		err = deleteMetric(os.Stdout, "projects/"+projectID+"/metricDescriptors/"+oldestType)
-		if err != nil {
-			panic(err.Error())
-		}
-		return
+		klog.Fatal(err.Error())
 	}
 
-	if publish != nil && *publish == true {
-		_, err = createCustomMetric(ctx, mc, os.Stdout, statusName, statusType, "The number of Workflow Runs with a status of `pending`")
-		if err != nil {
-			panic(err.Error())
-		}
-
-		_, err = createCustomMetric(ctx, mc, os.Stdout, oldestName, oldestType, "The number of seconds since the oldest Workflow Runs with a status of `pending` was started")
-		if err != nil {
-			panic(err.Error())
-		}
-	}
-
-	for {
-		statuses := getStatuses(ctx, kc)
-		count := 0
-		oldest := 0.0
-		for _, status := range statuses {
-			if status.Status == "pending" || status.Status == "" {
-				count += 1
-				if status.SecondsSinceStart > oldest {
-					oldest = status.SecondsSinceStart
-					fmt.Printf("Found older: %d\n", int(oldest))
-				}
-			}
-		}
-		statusSeries := makeTimeSeriesValue(statusType, *environment, count)
-		oldestSeries := makeTimeSeriesValue(oldestType, *environment, int(oldest))
-
-		if publish != nil && *publish == true {
-			err = writeTimeSeriesRequest(ctx, mc, []*monitoringpb.TimeSeries{statusSeries, oldestSeries})
+	go func() {
+		ticker := time.NewTicker(DefaultWorkflowRunPollingInterval)
+		defer ticker.Stop()
+		for ; true; <-ticker.C {
```

Comment:
I believe the most concise expression of this is `for range ticker.C { ... }`. You might also be able to express this as a `manager.Runnable` (https://github.com/kubernetes-sigs/controller-runtime/blob/v0.8.1/pkg/manager/manager.go#L249) that would give you a bit more control over the context situation.

Reply:
My goal was mostly a quick integration with the previous code, and in that case I wanted to avoid the infinite for loop and the sleep, both of which are pet peeves of mine. My intention was not to rewrite it entirely.

Reply:
A range over a channel isn't an infinite loop, but rather it ends when the channel is closed. Although in this case I think the `defer ticker.Stop()` outside of the loop will cause this to run forever anyway, although I may be misunderstanding the logic.

Reply:
Actually that's my point. I don't consider these for loops to be infinite.
The previous code was this:
```
	for {
                ...
		time.Sleep(INTERVAL * time.Second)
	}
```

Both of which are pet peeves of mine.